libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -712,71 +733,71 @@ void Compass::setHIL(uint8_t instance, float roll, float pitch, float yaw)
     }
 
     // convert the earth frame magnetic vector to body frame, and
     // apply the offsets
-    _hil.field[instance] = R.mul_transpose(_hil.Bearth);
+    _s_instance._hil.field[instance] = R.mul_transpose(_s_instance._hil.Bearth);
 
     // apply default board orientation for this compass type. This is
     // a noop on most boards
-    _hil.field[instance].rotate(MAG_BOARD_ORIENTATION);
+    _s_instance._hil.field[instance].rotate(MAG_BOARD_ORIENTATION);
 
     // add user selectable orientation
-    _hil.field[instance].rotate((enum Rotation)_state[0].orientation.get());
+    _s_instance._hil.field[instance].rotate((enum Rotation)_s_instance._state[0].orientation.get());
 
-    if (!_state[0].external) {
+    if (!_s_instance._state[0].external) {
         // and add in AHRS_ORIENTATION setting if not an external compass
-        _hil.field[instance].rotate(_board_orientation);
+        _s_instance._hil.field[instance].rotate(_s_instance._board_orientation);
     }
-    _hil.healthy[instance] = true;
+    _s_instance._hil.healthy[instance] = true;
 }
 
 // Update raw magnetometer values from HIL mag vector
 //
 void Compass::setHIL(uint8_t instance, const Vector3f &mag)
 {
-    _hil.field[instance] = mag;
-    _hil.healthy[instance] = true;
-    _state[instance].last_update_usec = AP_HAL::micros();
+    _s_instance._hil.field[instance] = mag;
+    _s_instance._hil.healthy[instance] = true;
+    _s_instance._state[instance].last_update_usec = AP_HAL::micros();
 }
 
-const Vector3f& Compass::getHIL(uint8_t instance) const
+const Vector3f& Compass::getHIL(uint8_t instance)
 {
-    return _hil.field[instance];
+    return _s_instance._hil.field[instance];
 }
 
 // setup _Bearth
 void Compass::_setup_earth_field(void)
 {
     // assume a earth field strength of 400
-    _hil.Bearth(400, 0, 0);
+    _s_instance._hil.Bearth(400, 0, 0);
 
     // rotate _Bearth for inclination and declination. -66 degrees
     // is the inclination in Canberra, Australia
     Matrix3f R;
     R.from_euler(0, ToRad(66), get_declination());
-    _hil.Bearth = R * _hil.Bearth;
+    _s_instance._hil.Bearth = R * _s_instance._hil.Bearth;
 }
 
 /*
   set the type of motor compensation to use
  */
 void Compass::motor_compensation_type(const uint8_t comp_type)
 {
-    if (_motor_comp_type <= AP_COMPASS_MOT_COMP_CURRENT && _motor_comp_type != (int8_t)comp_type) {
-        _motor_comp_type = (int8_t)comp_type;
-        _thr_or_curr = 0;                               // set current current or throttle to zero
+    if (_s_instance._motor_comp_type <= AP_COMPASS_MOT_COMP_CURRENT && _s_instance._motor_comp_type != (int8_t)comp_type) {
+        _s_instance._motor_comp_type = (int8_t)comp_type;
+        _s_instance._thr_or_curr = 0;                               // set current current or throttle to zero
         for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
             set_motor_compensation(i, Vector3f(0,0,0)); // clear out invalid compensation vectors
```

Comment:
this is a weird singleton implementation. What needs to be done is guaranteeing there's ever only 1 instance of this class. Then you don't need to prefix every member with _s_instance. You are already using the only possible instance. This usually means making the constructor private.
See how we do in AP_InertialSensor.